nilearn/connectivity/connectivity_matrices.py

```diff
@@ -0,0 +1,341 @@
+
+    if prop == 'spd':
+        if not is_spd(matrix, decimal=7):
+            raise ValueError('Expected a symmetric positive definite matrix.')
+
+
+def _form_symmetric(function, eigen_values, eigen_vectors):
+    """Return the symmetric matrix with eigenvectors eigen_vectors and
+    eigenvalues obtained by applying the function to eigen_values.
+
+    Parameters
+    ----------
+    function : function numpy.ndarray -> numpy.ndarray
+        The transform to apply to the eigenvalues.
+
+    eigen_values : numpy.ndarray, shape (n_features, )
+        Input argument of the function.
+
+    eigen_vectors : numpy.ndarray, shape (n_features, n_features)
+        Unitary matrix.
+
+    Returns
+    -------
+    output : numpy.ndarray, shape (n_features, n_features)
+        The symmetric matrix obtained after transforming the eigenvalues, while
+        keeping the same eigenvectors.
+    """
+    return np.dot(eigen_vectors * function(eigen_values), eigen_vectors.T)
+
+
+def _map_eigenvalues(function, symmetric):
+    """Matrix function, for real symmetric matrices. The function is applied
+    to the eigenvalues of symmetric.
+
+    Parameters
+    ----------
+    function : function numpy.ndarray -> numpy.ndarray
+        The transform to apply to the eigenvalues.
+
+    symmetric : numpy.ndarray, shape (n_features, n_features)
+        The input symmetric matrix.
+
+    Returns
+    -------
+    output : numpy.ndarray, shape (n_features, n_features)
+        The new symmetric matrix obtained after transforming the eigenvalues,
+        while keeping the same eigenvectors.
+
+    Note
+    ----
+    If input matrix is not real symmetric, no error is reported but result will
+    be wrong.
+    """
+    eigen_values, eigen_vectors = linalg.eigh(symmetric)
+    return _form_symmetric(function, eigen_values, eigen_vectors)
+
+
+def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
+    """Compute the geometric mean of symmetric positive definite matrices.
+
+    The geometric mean of n positive definite matrices
+    M_1, ..., M_n is the minimizer of the sum of squared distances from an
+    arbitrary matrix to each input matrix M_k
+
+    gmean(M_1, ..., M_n) = argmin_X sum_{k=1}^N dist(X, M_k)^2
+
+    where the used distance is related to matrices logarithm
+
+    dist(X, M_k) = ||log(X^{-1/2} M_k X^{-1/2)}||
+
+    In case of positive numbers, this mean is the usual geometric mean.
+
+    See Algorithm 3 of:
+        P. Thomas Fletcher, Sarang Joshi. Riemannian Geometry for the
+        Statistical Analysis of Diffusion Tensor Data. Signal Processing, 2007.
+
+    Parameters
+    ----------
+    matrices : list of numpy.ndarray, shape of each (n_features, n_features)
+        List of matrices whose geometric mean to compute. Raise an error if the
+        arrays are not all symmetric positive definite of the same shape.
+
+    init : numpy.ndarray, shape (n_features, n_features) or None, optional
```

Comment:
I find the or None a bit misleading (I originally thought the "or None" applied to the shape ...). Maybe just remove it, since you have optional and you explain what happens if init=None below.